src/package1/CustomArrayList.java

```diff
@@ -112,9 +112,14 @@ public boolean addAll(int index, Collection<? extends String> c) {
             String[] a = Arrays.copyOf(elements, (int) (elements.length + c.size() * 1.5));
             elements = a;
         }
-
-        System.arraycopy(elements, index, c, 0, c.size());
-        size=+c.size();
+        String[] newElements = new String[elements.length];
+        System.arraycopy(elements, 0, newElements, 0, index);
+        System.arraycopy(elements, index, newElements,
+                index + c.size(), size - index);
+        Object[] arrayC = c.toArray();
+        System.arraycopy(arrayC, 0, newElements, index, c.size());
+        elements = newElements;
+        size = size + c.size();
```

Comment:
I can't figure out the logic. Please add some comment to describe it

Reply:
1. Create new array 'newElement' because when copying to itsef there may be  overlapping elements depending on the length of both collections  and the value 'index'.
2. Copied elements that do not change position from 'elements' to 'newElements.'
3. Copied other elements, leaving empty cells for new elements.
4. Transformed collection 'c' to array.
5. Copied new elements into 'newElements'.
6. Increased 'size'.